Code readability improvements/optimizations

n-ski opened the pull request:
The list of changes (non-exhaustive but I tried to include everything):
- `is` and `as` patterns are used for safe casts
- Conditional statements now replace if/else branches for simple assignments/returns
- Conditional statements now check for non-`null` value rather than `null` value
- `Array.Empty<T>()` statement replaces `new T[0]` statement for creating zero-length arrays 
- Strings are checked for emptiness by comparing its `Length` property with `0`
- `readonly` modifiers were added to some fields
- Many instances of manually populating lists were replaces with simple LINQ `Select` queries (mainly when parsing `JToken` objects)
- Various pieces of code are now spaced out based on their purpose so code is less dense in general
- `JsonConvert.DeserializeObject<T>()` is preferred over doing the same with the `(T)JsonConvert.DeserializeObject()`
- User agent string (`"Mozilla/5.0 BooruSharp"`) is made into a constant in `ABooru` class
- Various argument exceptions are now passed the parameter name that caused it
- Got rid of regular expression in `Template.Gelbooru` class that was used to parse date/time string, it is now parsed using raw date/time format string with `DateTime.ParseExact()`
- Replaced `protected internal` access modifiers on some methods with `private protected` access modifiers. See commit message for details.
- Make sure HTTP response messages are successful after sending requests. This will help with random null reference exceptions and unexpected results in unit tests.

This is a rather large PR, so take your time to review it and point out inconsistencies and things you don't like/would want to see changed.

n-ski committed:
Improve code style
- Use patterns for safe casts
- When possible, use conditional statements for assignments/returns instead of doing the same with if/else branches
- When using conditional statements and checking for null, check for non-null value instead
- Use Array.Empty<T>() language construct for referencing empty arrays, don't create 0-length arrays manually
- When checking if the string is empty, check that its length is 0, don't compare it with an empty string
- Add readonly modifiers to some fields
- Replace many instances of manually populating lists with simple LINQ queries
- Space out various pieces of code based on their purpose
- Prefer generic JsonConvert.DeserializeObject<T> method over non-generic JsonConvert.DeserializeObject + cast

n-ski committed:
Improve ArgumentExceptions
- Feed ArgumentNullException constructor the parameter name, not the message, constructor will create the correct message for us
- Provide parameter name in ArgumentException constructors

n-ski committed:
Gelbooru: don't use regex to parse post creation date

n-ski commented on BooruSharp/Booru/Template/Gelbooru.cs:
```diff
@@ -116,7 +90,7 @@ protected internal override Search.Comment.SearchResult GetCommentSearchResult(o
                 int.Parse(elem.Attributes.GetNamedItem("id").Value),
                 int.Parse(elem.Attributes.GetNamedItem("post_id").Value),
-                creatorId.InnerText == "" ? (int?)null : int.Parse(creatorId.Value),
+                creatorId.InnerText.Length > 0 ? int.Parse(creatorId.Value) : (int?)null,
```
Not sure why `creatorId.InnerText` was used for the check, but `creatorId.Value` was used as the actual value to parse, so I left this as is.

n-ski commented on BooruSharp.Others/Pixiv.cs:
```diff
@@ -107,123 +121,170 @@ public override async Task AddFavoriteAsync(int postId)
-        private string _clientID = "MOBrBDS8blbauoSck0ZfDbtuzpyT";
-        private string _clientSecret = "lsACyCD94FhDUtGTXi3QzcFE2uU1hqtDaKeqrdwj";
-        private string _hashSecret = "28c1fdd170a5204386cb1313c7077b34f83e4aaf4aa829ce78c231e05b0bae2c";
+        private readonly string _clientID = "MOBrBDS8blbauoSck0ZfDbtuzpyT";
+        private readonly string _clientSecret = "lsACyCD94FhDUtGTXi3QzcFE2uU1hqtDaKeqrdwj";
+        private readonly string _hashSecret = "28c1fdd170a5204386cb1313c7077b34f83e4aaf4aa829ce78c231e05b0bae2c";
```
These can probably be made into `const string`s instead.

n-ski commented on BooruSharp/Booru/ABooru.cs:
```diff
@@ -222,23 +218,23 @@ private async Task<string> GetRandomIdAsync(string tags)
         private string TagsToString(string[] tags)
         {
-            return tags == null ? "" : "tags=" + string.Join("+", tags.Select(x => Uri.EscapeDataString(x))).ToLower();
+            return tags != null
+                ? "tags=" + string.Join("+", tags.Select(Uri.EscapeDataString)).ToLower()
+                : "";
```
This should probably also check for `tags.Length > 0`.

Xwilarg replied:
Since it's called internally I'm not sure checking the length is really important
(And iirc the public functions already remove empty tags)

n-ski commented on BooruSharp/Booru/ABooru.cs:
```diff
@@ -301,10 +294,11 @@ protected get
         protected static readonly Random _random = new Random();
         // TODO: remove this message after removing obsolete constructors.
         protected const string _deprecationMessage = "Use a contructor that accepts single BooruOptions parameter. Use | (bitwise OR) operator to combine multiple options.";
+        private const string _userAgentHeaderValue = "Mozilla/5.0 BooruSharp";
```
This should be replaced with something sensible like `"BooruSharp/1.2.3"` in the future (I have some logic that takes care of pulling assembly version in mind), at least according to [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent#Syntax). After all, we aren't using anything Mozilla related in this project.

Xwilarg replied:
I think we can change that without any risk of breaking things

Xwilarg commented on BooruSharp/Booru/ABooru.cs:
```diff
@@ -152,8 +152,6 @@ protected ABooru(string baseUrl, UrlFormat format, BooruOptions options)
                 _imageUrlXml = _imageUrl.Replace("json=1", "json=0");
             else if (_format == UrlFormat.postIndexJson)
                 _imageUrlXml = _imageUrl.Replace("index.json", "index.xml");
-            else
```
Why did you remove the else case?

n-ski replied:
IMO it's redundant, since by default `_imageUrlXml` will be initialized to `null`. The same reason why `_wikiUrl`, `_relatedUrl` and `_commentUrl` below aren't initialized with `null` explicitly.

Xwilarg commented on BooruSharp.Others/Pixiv.cs:
```diff
@@ -25,33 +25,40 @@ public class Pixiv : ABooru
-                var bytes = md5.ComputeHash(Encoding.UTF8.GetBytes(time + _hashSecret));
-                StringBuilder builder = new StringBuilder();
-                for (int i = 0; i < bytes.Length; i++)
-                    builder.Append(bytes[i].ToString("x2"));
-                request.Headers.Add("X-Client-Hash", builder.ToString());
+                var hashBytes = md5.ComputeHash(Encoding.UTF8.GetBytes(time + _hashSecret));
+                var hashString = BitConverter.ToString(hashBytes).Replace("-", null).ToLower();
+
+                request.Headers.Add("X-Client-Hash", hashString);
```
Could you please make sure these lines aren't the reason the Pixiv unit tests are failing?

n-ski replied:
Hex string output is the same. See https://dotnetfiddle.net/17PSj2

Xwilarg commented:
I'm actually not sure if we should fix the unit tests before approuving more PR, what is your opinion on that?
(Same question for #11)

n-ski committed:
Fix 'protected internal' access modifiers
The intention behind 'protected internal' was to make some methods only accessible by the subclasses defined in BooruSharp, but in fact outside projects still have 'protected' access to these methods, and also made these methods accessible by any class in BooruSharp project. 'private protected' actually fullfills that initial requirement.

n-ski committed:
Ensure success status code
Boorus may respond with unsuccessful status codes and since we don't handle those at all (except authentication), sometimes it can cause undefined behavior (like null reference exception or unexpected SearchResult). Now HttpRequestException will be thrown if requests weren't successful.
